chain-impl-mockchain/src/block/version.rs

```diff
@@ -1,32 +1,100 @@
+use lazy_static::lazy_static;
 use num_derive::FromPrimitive;
 use num_traits::FromPrimitive;
+use std::collections::BTreeMap;
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct BlockVersion(pub(crate) u16);
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AnyBlockVersion {
+    Supported(BlockVersion),
+    Unsupported(u16),
+}
 
-impl BlockVersion {
-    pub const fn new(v: u16) -> Self {
-        BlockVersion(v)
+impl AnyBlockVersion {
+    pub fn try_into_block_version(self) -> Option<BlockVersion> {
+        match self {
+            AnyBlockVersion::Supported(version) => Some(version),
+            AnyBlockVersion::Unsupported(_) => None,
+        }
+    }
+}
+
+impl PartialEq<BlockVersion> for AnyBlockVersion {
+    fn eq(&self, other: &BlockVersion) -> bool {
+        match self {
+            AnyBlockVersion::Supported(version) => version == other,
+            AnyBlockVersion::Unsupported(_) => false,
+        }
+    }
+}
+
+impl From<u16> for AnyBlockVersion {
+    fn from(n: u16) -> Self {
+        match BlockVersion::from_u16(n) {
+            Some(supported) => AnyBlockVersion::Supported(supported),
+            None => AnyBlockVersion::Unsupported(n),
+        }
     }
+}
+
+impl Into<u16> for AnyBlockVersion {
+    fn into(self) -> u16 {
+        match self {
+            AnyBlockVersion::Supported(version) => version as u16,
+            AnyBlockVersion::Unsupported(n) => n,
+        }
+    }
+}
 
-    pub const fn as_u16(&self) -> u16 {
-        self.0
+impl From<BlockVersion> for AnyBlockVersion {
+    fn from(version: BlockVersion) -> Self {
+        AnyBlockVersion::Supported(version)
     }
 }
 
 #[derive(Debug, Clone, Copy, FromPrimitive, PartialEq, Eq)]
-pub enum BlockVersionTag {
-    ConsensusNone = 0x0,
-    ConsensusBft = 0x1,
-    ConsensusGenesisPraos = 0x2,
+pub enum BlockVersion {
+    Genesis = 0,
+    Ed25519Signed = 1,
+    KesVrfproof = 2,
 }
 
-impl BlockVersionTag {
-    pub fn to_block_version(self) -> BlockVersion {
-        BlockVersion::new(self as u16)
+impl BlockVersion {
+    pub fn get_consensus(self) -> ConsensusVersion {
+        match self {
+            BlockVersion::Genesis => ConsensusVersion::None,
+            BlockVersion::Ed25519Signed => ConsensusVersion::Bft,
+            BlockVersion::KesVrfproof => ConsensusVersion::GenesisPraos,
+        }
     }
+}
+
+#[derive(Debug, Clone, Copy, FromPrimitive, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ConsensusVersion {
+    None = 0,
+    Bft = 1,
+    GenesisPraos = 2,
+}
 
-    pub fn from_block_version(ver: BlockVersion) -> Option<BlockVersionTag> {
-        BlockVersionTag::from_u16(ver.0)
+impl ConsensusVersion {
+    pub fn supported_block_versions(self) -> &'static [BlockVersion] {
```

Comment:
whilst saving potential mismatch if you update one without the other, that seems a bit overkill in this case